src/middlewared/middlewared/plugins/pool.py

```diff
@@ -4182,6 +4184,148 @@ def __handle_zfs_set_property_error(self, e, properties_definitions):
+
+            except Exception:
+                self.logger.debug('Unable to resolve uid %d to name', entry['uid'])
+                pass
+
+            quota_list.append(entry)
+
+        return filter_list(quota_list, filters, options)
+
+    @accepts(
+        Str('id', required=True),
+        Ref('query-filters'),
+        Ref('query-options'),
+    )
+    @item_method
+    async def get_groupquota(self, id, filters, options):
+        """
+        Returns a list of group quotas on the ZFS dataset `id`. Supports
+        `query-filters` and `query-options`.
+        """
+        quota_list = []
+        groupquota = await run(
+            ['zfs', 'groupspace', '-H', '-n', '-p', '-o', 'name,used,quota', id],
+            check=False
+        )
+        if groupquota.returncode != 0:
+            raise CallError('Failed to get groupquota for %s: [%s]',
+                            id, groupquota.stderr.decode())
+
+        for quota in groupquota.stdout.decode().splitlines():
+            entry = {
+                'gid': None,
+                'groupname': None,
+                'quota': 0,
+                'used': 0,
+            }
+            m = RE_GROUPSPACE.match(quota)
+            if m is None:
+                self.logger.debug('Invalid groupspace quota: %s', quota)
+                continue
+
+            entry.update(m.groupdict())
+
+            try:
+                entry['groupname'] = (
+                    await self.middleware.call('group.get_group_obj', {'gid': entry['gid']})
+                )['gr_name']
+
+            except Exception:
+                self.logger.debug('Unable to resolve gid %d to name', entry['gid'])
+                pass
+
+            quota_list.append(entry)
+
+        return filter_list(quota_list, filters, options)
+
+    @accepts(
+        Str('id', required=True),
+        List('quotas', items=[
+            Dict(
+                'userspace_quota',
+                Str('type', enum=['USER', 'GROUP'], required=True),
+                Str('id', required=True),
+                Int('quota', required=True),
+                Str('unit', enum=['B', 'KB', 'MB', 'GB', 'TB'], default='B')
+            )
+        ])
+    )
+    @item_method
+    async def set_userspace_quota(self, id, data):
+        """
+        Set userspace (user or group) quotas on the specified ZFS dataset.
+
+        `id` name of ZFS dataset
+
+        `quotas` list of `userspace_quota` to apply to dataset.
+
+        userspace_quota entries have the following parameters:
+
+        `type`: whether it's a user or group quota.
+
+        `id`: the uid or gid to which the quota applies.
+
+        `quota`: the quota size.
+
+        `unit`: the units expressed in the quota size (defaults to bytes).
+        """
+        units = {
+            'TB': 1099511627776,
+            'GB': 1073741824,
+            'MB': 1048576,
+            'KB': 1024,
+            'B': 1
+        }
+        cmd = ['zfs', 'set']
+        for q in data:
```

Comment:
Probably need to add validation for length of list. May want to set upper-bound on how many quotas to set in a single command.

Reply:
I agree